### ISSUE TITLE
not getting same elevation and azimuth as NOAA

### ISSUE THREAD

**Issue by dlarue** [jpb10/SolarCalculator#3]
Hi, I've made some simple ports of this to Linux which are mostly printf instead of Serial.print and adding <math.h> and <stdio.h> but what I'm getting is off from what NOAA states for San Diego CA. https://gml.noaa.gov/grad/solcalc/

NOAA would say my solar elevation is 43.61 but the library calculation gives me 19.7
NOAA would say my solar azimuth is 236.3 but the library calculation gives me 100.6

The calculated sunrise and sunset are off by 2 minutes and 3 minutes compared to NOAA so pretty close.

I'm using Lat long in the library as obtained from the NOAA web page when the marker is placed at the airport on the road to Harbor Island Dr:
double latitude = 32.727219;
double longitude = -117.19665;

Any ideas on what could be causing this elevation and azimuth discrepancy?

**Comment by dlarue** [comment]
Pasted below is my main.cpp and this is what I use to compile: gcc main.cpp SolarCalculator.cpp -o main -lm

`//======================================================================================================================
// SolarCalculator Library for Arduino example sketch: SunriseSunsetAltitude.ino
//
// Calculate the rise and set times at a height above the level of the horizon.
//
// Tested with Arduino IDE 1.8.19 and Arduino Uno
//======================================================================================================================

#include "SolarCalculator.h"
#include <math.h>
#include <time.h>
#include <stdio.h>

// Rounded HH:mm format
///*
char * hoursToString(double h, char *str)
{
  int m = int(round(h * 60));
  int hr = (m / 60) % 24;
  int mn = m % 60;

  str[0] = (hr / 10) % 10 + '0';
  str[1] = (hr % 10) + '0';
  str[2] = ':';
  str[3] = (mn / 10) % 10 + '0';
  str[4] = (mn % 10) + '0';
  str[5] = '\0';
  return str;
}
//*/

int main()
{
  //Serial.begin(9600);

  // Date
  time_t now;
  struct tm * timeinfo;
  time(&now);
  timeinfo = localtime(&now);

  int hours;
  int seconds;
  int minutes;
  int days;
  int months;
  int years;

  years = timeinfo->tm_year + 1900;   //https://mikaelpatel.github.io/Arduino-RTC/d8/d5a/structtm.html
  months = timeinfo->tm_mon + 1;
  days = timeinfo->tm_mday;
  hours = timeinfo->tm_hour;
  minutes = timeinfo->tm_min;
  seconds = timeinfo->tm_sec;


  // Location
  double latitude = 32.727219;
  double longitude = -117.19665;
  int utc_offset = -7; // used for finding sunrise and sunset

  double transit, sunrise, sunset;

  // From the Explanatory Supplement to the Astronomical Almanac (1992), p. 484
  // Sunrise or sunset at a height above the level of the horizon occurs when the Sun's altitude is approximately:

  int height = 200;  // in meters
  double sun_altitude = SUNRISESET_STD_ALTITUDE - 0.0353 * sqrt(height);

  // Calculate the times of sunrise, transit, and sunset, in hours (UTC)
  calcSunriseSunset(years, months, days, latitude, longitude, transit, sunrise, sunset, sun_altitude);

  // Print results
  char str[6];
  printf("SunRise:");
  printf("%s\n",hoursToString(sunrise + utc_offset, str));
  printf("Transit:");
  printf("%s\n",hoursToString(transit + utc_offset, str));
  printf("SunSet:");
  printf("%s\n",hoursToString(sunset + utc_offset, str));

  double sunElevation;
  double sunAzimuth;

  // Calculate the Sun's azimuth and elevation (corrected for atmospheric refraction), in degrees
  //calcHorizontalCoordinates(year(), month(), day(), hour(), minute(), second(), latitude, longitude, sunAzimuth, sunElevation);
  calcHorizontalCoordinates(years, months, days, hours, minutes, seconds, latitude, longitude, sunAzimuth, sunElevation);

  printf("Day: %d: %d: %d\n",months,days,years);
  printf("Time: %d: %d: %d\n",hours,minutes,seconds);
  printf("Sun angles : \n");
  printf("    Elevation : ");
  printf("%f",sunElevation);
  printf("    panel Elevation : ");
  printf("%f\n",90 - sunElevation);
  printf("    Azimuth   : ");
  printf("%f\n",sunAzimuth);
}


`

**Comment by jpb10** [comment]
You are probably using local time instead of UTC. 

If local time is 3:00 PM (San Diego CA), then UTC is 10:00 PM.

**Comment by dlarue** [comment]
Ah, that would make sense and yes I'm using local time. ugh. Thank you!

**Comment by jpb10** [comment]
No problem. Also the sunrise and sunset times should be the same as NOAA. 

Check if you are using the correct UTC day.

**Comment by dlarue** [comment]
I did some work with a NOAA tide chart library and should have remembered they use UTC times...
I'm using tz of -7 so it should have been the same UTC day. My generated times were off by 2 minutes at sunrise and 3 minutes for sunset.  I will check on that again once I get my code cleaned up for handling both local and UTC time variables.

**Comment by dlarue** [comment]
I just used a few new variables to get a timeinfo structure with UTC data and now I get the same elevation and azimuth as the NOAA site when using the UTC time/date. 

` utc_now = time( NULL );
  timeinfo_utc = gmtime(&utc_now);
`
But my sunrise and sunset calculations are the same, ie 2 minutes off at sunrise and 3 minutes for sunset. While transit matches NOAA.

**Comment by jpb10** [comment]
```
int height = 200; // in meters
double sun_altitude = SUNRISESET_STD_ALTITUDE - 0.0353 * sqrt(height);

// Calculate the times of sunrise, transit, and sunset, in hours (UTC)
calcSunriseSunset(years, months, days, latitude, longitude, transit, sunrise, sunset, sun_altitude);
```

In the code above, you're correcting for the dip in the horizon of an observer at a height of 200 m above the horizon. 

It's expected that the sunrise will be a few minutes early and the sunset a few minutes late versus the uncorrected times. 

NOAA doesn't correct for dip of the horizon. So to get the same times, remove the `sun_altitude` parameter.

**Comment by dlarue** [comment]
That makes sense and adjusting that to ~10 m altitude brings the numbers closer. I also see on the NOAA site they are using apparent sunset/rise which as you mentioned is a bit earlier/later than actual sunrise/set. 

When I started looking at this a few hours ago the sun was pretty high( ~2p ) and a guess was that it was up at 60 to 70 degrees but even then the NOAA site was saying it was 40 degrees which is pretty low. That number is also easy to estimate since 45 deg is half of 90 and it was well above 45 degrees above the horizon.  The library matches NOAA but what would explain the low elevation value?

**Comment by jpb10** [comment]
We are nearing the fall equinox (solar declination of 0 deg) and at a latitude of 32.7 deg, the solar elevation is 90 - 32.7 = 57.3 deg at transit on Sep 22, 2022.

In other words, 60 to 70 degrees is impossible. 😉

**Comment by dlarue** [comment]
Tomorrow I will get a compass and a level out and rig up a way to measure the sun angle and compare it to the results.
I just realized I have an app on my phone which I could have used.... I'll be pleased to be wrong.

I'm considering making a little desktop 'toy' which will track the sun as it moves overhead with a couple of servos. There's a servo 'easing' library I'd like to try and and this would make for a fun one.

**Comment by jpb10** [comment]
I'm closing this issue, but don't hesitate to post your measurements.

**Comment by dlarue** [comment]
So I used a compass with a 12" long rod attached to the pivot of the compass and leveled the compass on a "bullet" level. Then I raised the rod to point to the sun, rotated the unit at the same time and observed the rod's shadow on my fingers as I raised and pointed it until the shadow was unseen. When I snapped the picture the level bubble was slightly off level and the pointer was just about pointing to 55 or 56 degrees.  The library told me the current elevation is 54.2 degrees so it's a good output!

Previously, when I first got this running on my Linux laptop, it was around this same time and because I was using Local Time instead of UTC, it was telling me the elevation was close to 40 degrees. I'll add my main.cpp to my fork and document building in the main.cpp file.
